fdbserver/DDRelocationQueue.actor.cpp

```diff
@@ -530,6 +530,35 @@ struct DDQueue : public IDDRelocationQueue {
 		Future<Void> cancel;
 	};
 
+	struct MovingAverage {
+		int64_t previous;
+		int64_t total;
+		double interval;
+		Deque<std::pair<double, int64_t>> updates; // std::pair<time, bytes>
+
+		int64_t getTotal() { return total; }
+
+		double getAverage() {
+			while (!updates.empty() && updates.front().first < now() - interval) {
+				previous += updates.front().second;
+				updates.pop_front();
+			}
+			return (total - previous) / interval;
+		}
+
+		void sumBytes(int64_t bytes) {
+			total += bytes;
+			updates.push_back(std::make_pair(now(), bytes));
```

Comment:
There should be some check here on the size of `updates` so that we don't OOM the server. (It's unlikely to happen, but if it happens, that could be pretty bad.) Something like: if `updates` is greater than a particular size, then pop from the front until we're at size (and update `previous` accordingly).

Reply:
The `getAverage()` function will have change accordingly since it will be possible that the first update in `updates` is less than `interval` time ago, so `total - previous` should be divided by the diff in timestamps.